design-documents/frontend/technical-vision.md

```diff
@@ -0,0 +1,140 @@
+   - Logging with [redux-logger](https://github.com/evgenyrodionov/redux-logger).
+- **Internationalization** with [react-intl](https://www.npmjs.com/package/react-intl).
+   - _Disclaimer: the documented path here is to couple the components to this library. This will be fine for most projects, but if ultimate **portability** is required, you may want to decouple with the [container paradigm][]._
+
+## [TypeScript][]
+
+_Maintainability, Productivity, Quality, Consistency._
+
+[TypeScript][] is not going to solve all of our problems, but its static typing features definitely have the potential to improve the **quality** of the code we produce. [TypeScript][] code is also easier to **maintain**, because it purposefully makes it more difficult to inadvertently break something that was already defined to work a certain way. As a superset of JavaScript, it allows us to leverage both the [JavaScript ecosystem](https://www.npmjs.com/) as well as any existing developer knowledge. It also happens to [work quite well with React and JSX syntax](https://www.typescriptlang.org/docs/handbook/jsx.html), replacing [prop-types](https://www.npmjs.com/package/prop-types) (run-time) with [TypeScript Interfaces](https://www.typescriptlang.org/docs/handbook/interfaces.html) (compile-time).
+
+> Types and interfaces are only a compile-time construct and have no effect on the generated code.
+
+Catching errors at compile time also increases **productivity**, because we don't have to run the development server to see the errors. In fact, we can see them directly in the editor before anything is run. Additionally, Microsoft has open sourced the [TypeScript Language Service API](https://github.com/Microsoft/TypeScript/wiki/Using-the-Language-Service-API), which is used across the most popular editors of today. As a result, you'll see **consistent** errors between these editors.
+
+[Type definitions](https://www.typescriptlang.org/docs/handbook/declaration-files/consumption.html) are provided for [many of the most popular npm packages](http://definitelytyped.org/), meaning developers can increase **productivity** by staying in-editor instead of constantly task switching and looking up documentation online. Even internally, there is less need to open a local file when you can extract most of the relevant information you need via [Intellisense](https://code.visualstudio.com/docs/editor/intellisense#_intellisense-features).
+
+_Disclaimer: though possible to introduce [TypeScript][] into an existing [CRA][] app and though [TypeScript][] can co-exist with JavaScript, it may not make sense to introduce it into an existing project if you feel it may hinder momentum._
+
+### Technologies
+- [Adding TypeScript to CRA](https://facebook.github.io/create-react-app/docs/adding-typescript).
+- Code **consistency** is enforced via [configuration](http://www.typescriptlang.org/docs/handbook/tsconfig-json.html) and [linting with TSLint](https://palantir.github.io/tslint/).

```

Comment:
Can we add in a section for Prettier and be more clear about what category of rules we expect to be enabled with TSLint? We do too much bikeshedding on formatting in existing projects and I'd like to leave that as a relic of the past

Reply:
Yep. I need to do a bit more research here as to getting TSLint and Prettier and EditorConfig to work well together. All of this will change, of course, once [these TypeScript roadmap items](https://github.com/Microsoft/TypeScript/issues/29288#roadmap-overview) are addressed.

Reply:
Addressed in 4f625a47b50e0c0ac67ccbf5fed03eff17b94d40